Firestore/Source/Remote/FSTDatastore.m

```diff
@@ -557,21 +614,82 @@ - (void)resumeStartFromBackoff {
 
   // Momentarily set state to FSTStreamStateInitial as `start` expects it.
   self.state = FSTStreamStateInitial;
-  [self start];
+  [self startWithDelegate:delegate];
   FSTAssert([self isStarted], @"Stream should have started.");
 }
 
-- (void)stop {
-  FSTLog(@"%@ %p stop", NSStringFromClass([self class]), (__bridge void *)self);
+/**
+ * Closes the stream and cleans up as necessary:
+ *
+ * <ul>
+ *   <li>closes the underlying GRPC stream;
+ *   <li>calls the onClose handler with the given 'error';
+ *   <li>sets internal stream state to 'finalState';
+ *   <li>adjusts the backoff timer based on the error
+ * </ul>
+ *
+ * A new stream can be opened by calling `start` unless `finalState` is set to
+ * `FSTStreamStateStopped`.
+ *
+ * @param finalState the intended state of the stream after closing.
+ * @param error the NSError the connection was closed with.
+ */
+- (void)closeWithFinalState:(FSTStreamState)finalState error:(nullable NSError *)error {
+  FSTAssert(finalState == FSTStreamStateError || error == nil,
```

Comment:
This allows the error to be nil if the finalState is FSTStreamStateError. Is that desirable?

Reply:
This is directly called from GRPC and we don't want to crash the client here if the server cleanly closes our connection for some unknown reason.